CHANGES

```diff
@@ -19,6 +19,9 @@ Incompatible changes
 * #6830: py domain: ``meta`` fields in info-field-list becomes reserved.  They
   are not displayed on output document now
 * The structure of ``sphinx.events.EventManager.listeners`` has changed
+* Due to the scoping changes for :rst:dir:`productionlist` some uses of
+  :rst:role:`token` must be modified to include the scope which was previously
+  ignored.
```

Comment:
I feel this is too big incompatible change. How about adding warnings?

On my local, this works fine.
```
diff --git a/sphinx/domains/std.py b/sphinx/domains/std.py
index 725b879a9..6f61e877a 100644
--- a/sphinx/domains/std.py
+++ b/sphinx/domains/std.py
@@ -725,6 +725,8 @@ class StandardDomain(Domain):
                           RemovedInSphinx40Warning)
             domain = env.get_domain('citation')
             return domain.resolve_xref(env, fromdocname, builder, typ, target, node, contnode)
+        elif typ == 'token':
+            resolver = self._resolve_token_xref
         else:
             resolver = self._resolve_obj_xref

@@ -859,6 +861,27 @@ class StandardDomain(Domain):
         return make_refnode(builder, fromdocname, docname,
                             labelid, contnode)

+    def _resolve_token_xref(self, env: "BuildEnvironment", fromdocname: str,
+                            builder: "Builder", typ: str, target: str,
+                            node: pending_xref, contnode: Element) -> Element:
+        docname, labelid = '', ''
+        if ('token', target) in self.objects:
+            docname, labelid = self.objects['token', target]
+        elif ':' not in target:
+            # old styled reference
+            logger.warning('old styled token ref found. Please add a tokenGroup as a prefix',
+                           location=node)
+
+            target = ':' + target
+            for objtype, name in self.objects:
+                if objtype == 'token' and name.endswith(target):
+                    docname, labelid = self.objects[objtype, name]
+                    break
+
+        if not docname:
+            return None
+        return make_refnode(builder, fromdocname, docname, labelid, contnode)
+
     def _resolve_obj_xref(self, env: "BuildEnvironment", fromdocname: str,
                           builder: "Builder", typ: str, target: str,
                           node: pending_xref, contnode: Element) -> Element:
```

Reply:
In nitpicky-mode there will already be a warning about a reference not being resolved.
Note that the *productionGroup* argument is not becoming required with this PR so ``:token:`something` `` is still perfectly valid with, e.g.:
```rst
... productionlist::
   something: a b c
```
That means that as long as authors did not provide a *productionGroup* then there should be no extra warnings for those ``:token:`` roles.

Reply:
Thank you for explanation. Okay, go with this.